packages/next-auth/src/core/types.ts

```diff
@@ -313,30 +313,59 @@ export interface CallbacksOptions<P = Profile, A = Account> {
    * [`getSession`](https://next-auth.js.org/getting-started/client#getsession) |
    *
    */
-  session: (params: {
-    session: Session
-    user: User | AdapterUser
-    token: JWT
-  }) => Awaitable<Session>
+  session: (
+    params:
+      | { session: { user: DefaultJWT; expires: ISODateString }; token: JWT }
+      | { session: DefaultSession; user: AdapterUser; status: undefined }
+      | {
+          session: DefaultSession
+          user: AdapterUser
+          status: "update"
+          /**
+           * When using {@link SessionOptions.strategy} `"database"`, this is the data
+           * sent from the client via the [`useSession().update`](https://next-auth.js.org/getting-started/client#update-session) method.
+           *
+           * ⚠ Note, you should validate this data before using it.
+           */
+          newSession?: any
+        }
+  ) => Awaitable<Session>
   /**
    * This callback is called whenever a JSON Web Token is created (i.e. at sign in)
    * or updated (i.e whenever a session is accessed in the client).
    * Its content is forwarded to the `session` callback,
    * where you can control what should be returned to the client.
    * Anything else will be kept from your front-end.
    *
-   * ⚠ By default the JWT is signed, but not encrypted.
+   * The JWT is encrypted by default.
    *
    * [Documentation](https://next-auth.js.org/configuration/callbacks#jwt-callback) |
    * [`session` callback](https://next-auth.js.org/configuration/callbacks#session-callback)
    */
-  jwt: (params: {
-    token: JWT
-    user?: User | AdapterUser
-    account?: A | null
-    profile?: P
-    isNewUser?: boolean
-  }) => Awaitable<JWT>
+  jwt: (
+    params:
+      | {
+          token: DefaultJWT
+          user: User | AdapterUser
+          account: A | null
+          profile?: P
+          isNewUser?: boolean
+          status: "signIn"
+        }
+      | {
+          token: JWT
+          isNewUser?: boolean
+          /**
+           * When using {@link SessionOptions.strategy} `"jwt"`, this is the data
+           * sent from the client via the [`useSession().update`](https://next-auth.js.org/getting-started/client#update-session) method.
+           *
+           * ⚠ Note, you should validate this data before using it.
+           */
+          session?: any
+          status: "update"
+        }
+      | { token: JWT; isNewUser?: boolean }
```

Comment:
Hitting TypeScript issues, any help is appreciated:

![image](https://user-images.githubusercontent.com/18369201/227725760-53bb5692-8c4e-4e80-bcd7-53fd0cfa5053.png)

Reply:
You can remove the last `| { token: JWT; isNewUser?: boolean }` to make the type simpler and make typescript discriminated types work

To tell that `status` could be undefined use `status?: "update"`

Reply:
I would also deprecate the `isNewUser` field and add another status option with `signUp`

Reply:
We don't want to deprecate anything yet, this should be a non-breaking update. But agree, we can add `signUp`. :+1: